### ISSUE TITLE
Difference in DisallowSpaceIndentSniff between 3.1.1 and master

### ISSUE THREAD

**Issue by PatrickRose** [squizlabs/PHP_CodeSniffer#1772]
With the following file (tabs indentation), there's a difference between 3.1.1 and master

```php
<?php

class Example
{
	
	/**
	 *
	 */
	public function exampleFunction()
	{
		//
	}

}
```

In 3.1.1. you get:

```
$ phpcs --sniffs=Generic.WhiteSpace.DisallowSpaceIndent tmp.php

$
```

In master:
```
$ phpcs --sniffs=Generic.WhiteSpace.DisallowSpaceIndent tmp.php

FILE: /srv/www/zedcore/prose.trac-php7.zcode.net/tmp.php
---------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
---------------------------------------------------------------------------
 7 | ERROR | [x] Tabs must be used to indent lines; spaces are not allowed
 8 | ERROR | [x] Tabs must be used to indent lines; spaces are not allowed
---------------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
---------------------------------------------------------------------------

Time: 101ms; Memory: 8Mb

$
```

I think 3.1.1 acts correctly. I haven't bisected down to try and work out where it changed behaviour.

**Comment by jrfnl** [comment]
@PatrickRose That sniff needs to know what the `tab-width` is to work properly. Are you passing it a `--tab-with=#` on the command line or via a custom ruleset ?
If not, please do and the errors you are seeing should go away.

**Comment by PatrickRose** [comment]
I might be misremembering - didn't it use to default to a tab width of 4?

Either way, I don't think I should need to set a tab width - it should be
able to work out that it's for alignment because it's not followed by
another tab and the line doesn't begin with a space

**Comment by jrfnl** [comment]
> I might be misremembering - didn't it use to default to a tab width of 4?

@PatrickRose Yes, but that changed in PHPCS 3.0.0

> Either way, I don't think I should need to set a tab width - it should be able to work out that it's for alignment because it's not followed by another tab and the line doesn't begin with a space

The sniff allows for alignment, but only when alignment < tab-width.

**Comment by gsherwood** [comment]
Unfortunately, github has preserved the tab/space indents in the sample code, but I'm assuming you have a single space before the asterisk on line 7 and 8, and no spaces anywhere else.

If I use that sample code, I can replicate your error messages on master, but only when I don't set a tab-width.

The commit that changed the behaviour of this sniff looks to be this one: https://github.com/squizlabs/PHP_CodeSniffer/commit/eb3c55f93cedf3cb2aa14f8372105453d05c2068#diff-1475e4fabdbcf5646f594417dc7d9ba6

I haven't had time to look into why yet.

**Comment by gsherwood** [comment]
Ah, I see it. `$expectedIndentSize` is new, and is using the token length from the array. But if you aren't replacing tabs with spaces, then the length is the number of chars (counting a tab as one char) and that throws everything off.

**Comment by gsherwood** [comment]
To function properly, this sniff really needs tab width to be set on the CLI or in a standard. So if that isn't the case, it now asks the tokenizer to replace tabs in each token it has to check so it can still determine the correct token length.

If possible, I'd still recommend setting the tab width explicitly if you are using tabs for indentation. It's going to help out any sniffs that haven't been written with tab indents in mind.

Thanks a lot for testing master and reporting this issue. Can you have a try of the fix and see if it is working for you.

**Comment by PatrickRose** [comment]
I'll try today when I get in but I don't think that PHPCS should behave like that. Imagine I have some chained method calls that I want to indent. I would normally do

```php
$someVar->someMethod()
        ->nextMethod();
```

I've not verified it, but I think with the tab width set to 4 I'd get an error saying I should use tabs for indentation. In which case, I'll update it to

```php
$someVar->someMethod()
\t    ->nextMethod();
```

If someone has their tab widths set to display to something else (part of the reason I dislike tabs but it's the coding style at work so I live with it) then it won't be aligned correctly. Now obviously the coding style should say treat tab widths as 4, but someone else outside the project could open the code in their editor with a different tab width. 

In my opinion, DisallowSpaceIndent shouldn't try and work out whether the spaces are alignment or not by checking whether the spaces are less than the tabWidth because it doesn't know what I'm trying to align with.

**Comment by gsherwood** [comment]
@PatrickRose You're probably after this requested feature: #1586 

If this sniff completely ignored spaces after tabs, it wouldn't report errors for completely legitimate cases like:
```
if·($foo)·{
»   if·($bar)·{
»   ····echo "$foo $bar";
»   }
}
```
And for me, that means more bug reports because the sniff isn't actually doing what it is supposed to be doing.

What you need is for it to be smarter or configurable. But you probably don't want it to ignore cases like above in any case.

**Comment by PatrickRose** [comment]
@gsherwood - would it be more sensible to count the number of expected indentations (ie, the count of conditions for the first non-whitespace token on a line) instead of trying to do something fancy like that? That way, it can just go "I see there are two tabs followed by N spaces, I'm expecting to see two levels of indentation so that's fine"

Setting the tab widths via command line or in the ruleset works for now.

**Comment by gsherwood** [comment]
> would it be more sensible to count the number of expected indentations

That would require the sniff to essentially replicate the functionality of the Generic.WhiteSpace.ScopeIndent sniff, which is (by far) the most complex sniff included with PHPCS due to the huge number of indent cases it has to account for.

So yeah, it would probably be better to figure out the expected indent levels, but I suspect that wont work in all indent cases, and figuring out the indent is harder than you might think (again, for all cases).

**Comment by gsherwood** [comment]
I'm going to close this as it's blocking the 3.2.0 release and I think the original reported issue has been fixed. Thanks again for reporting it before it got released.
